duktape/src/main/jni/DuktapeContext.cpp

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright (C) 2016 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "DuktapeContext.h"
+#include "JString.h"
+#include "JavaMethod.h"
+
+namespace {
+
+// Internal names used to properties in the duktape context's global stash and bound variables.
+// The \xff\xff part keeps the variable hidden from JavaScript (visible through C API only).
+const char* JAVAVM_PROP_NAME = "\xff\xffjavaVM";
+const char* JAVA_THIS_PROP_NAME = "\xff\xffjava_this";
+const char* JAVA_METHOD_PROP_NAME = "\xff\xffjava_method";
+
+JNIEnv* getJNIEnv(duk_context *ctx) {
+  duk_push_global_stash(ctx);
+  duk_get_prop_string(ctx, -1, JAVAVM_PROP_NAME);
+  JavaVM* javaVM = static_cast<JavaVM*>(duk_require_pointer(ctx, -1));
+  duk_pop_2(ctx);
+  JNIEnv* env;
+  javaVM->AttachCurrentThread(&env, nullptr);
+  return env;
+}
+
+jobject getJavaThis(duk_context* ctx) {
+  duk_push_this(ctx);
+  duk_get_prop_string(ctx, -1, JAVA_THIS_PROP_NAME);
+  jobject thisObject = static_cast<jobject>(duk_require_pointer(ctx, -1));
+  duk_pop_2(ctx);
+  return thisObject;
+}
+
+JavaMethod* getJavaMethod(duk_context* ctx) {
+  duk_push_current_function(ctx);
+  duk_get_prop_string(ctx, -1, JAVA_METHOD_PROP_NAME);
+  JavaMethod* method = static_cast<JavaMethod*>(duk_require_pointer(ctx, -1));
+  duk_pop_2(ctx);
+  return method;
+}
+
+duk_int_t eval_string_with_filename(duk_context *ctx, const char *src,
+                                    const char *fileName) {
+  duk_push_string(ctx, fileName);
+  return duk_eval_raw(ctx, src, 0, DUK_COMPILE_EVAL | DUK_COMPILE_SAFE |
+                                   DUK_COMPILE_NOSOURCE | DUK_COMPILE_STRLEN);
+}
+
+duk_ret_t handler(duk_context *ctx) {
+  JavaMethod* method = getJavaMethod(ctx);
+  return method ? method->Invoke(ctx, getJNIEnv(ctx), getJavaThis(ctx)) : DUK_RET_INTERNAL_ERROR;
+}
+
+duk_ret_t javaObjectFinalizer(duk_context *ctx) {
+  jobject thisObject = static_cast<jobject>(duk_require_pointer(ctx, -1));
+  getJNIEnv(ctx)->DeleteGlobalRef(thisObject);
+  duk_enum(ctx, -1, DUK_ENUM_OWN_PROPERTIES_ONLY);
+  while (duk_next(ctx, -1, true)) {
+    if (duk_get_prop_string(ctx, -1, JAVA_METHOD_PROP_NAME)) {
+      delete static_cast<JavaMethod*>(duk_require_pointer(ctx, -1));
```

Comment:
Yes, it's supposed to be called when duktape's GC runs and there are no more JS references to our bound Java object.  It's hooked up below (see duk_set_finalizer).